### DISCUSSION TITLE
bevy_ecs - can the simd 'faster' crate be utilitized

### DISCUSSION THREAD

**Discussion by JeffGillin** [bevyengine/bevy#1822] [Ideas]
The performance of bevy_ecs is already impressive, but wondering if could make even faster with SIMD - the 'faster' crate can be used to get simd functionality. Don't know if this would be something made use of in application code (in ecs 'system' functions for example) versus integrated into ecs framework .

**Comment by bjorn3** [comment]
What exactly do you think SIMD can be used for? Many problem are not easily amendable to SIMD.

**Reply by JeffGillin** [reply]
Gotcha, thanks.  I'm only starting to learn about it and haven't dug into examples yet so don't understand the domain that solution applies to yet. But just thinking in general terms.  A system function is running on some thread with some set of data to process.  If that data can be processed faster using simd, then could get big improvements. But I guess that would be app-level code anyways, not framework.

**Reply by TheRawMeatball** [reply]
FYI, the glam crate used by bevy_math already uses SIMD under the hood.

**Reply by JeffGillin** [reply]
oh gotcha, thanks. that's great.

**Reply by InBetweenNames** [reply]
The ECS SoA representation is already in an ideal format for SIMD.  An example problem that comes up frequently and is amenable to SIMD is the calculation `position' = position + time*velocity`.  One would batch queries according to a multiple of the native vector width as in #1990 and then use SIMD operations (e.g., from the `faster` crate or something like C++'s `xsimd`) to perform calculations in parallel.  

It's worth mentioning as well that the AoSoA layout might be worth evaluating for the ECS if this problem area is explored more.

**Reply by InBetweenNames** [reply]
On reflecting on this a little more and inspecting the ECS code, it seems like my previous comment wasn't entirely accurate... A Vec3 component would still get stored with a layout similar to [x,y,z,x,y,z...] (or [x,y,z,1] if affine).  Currently, it seems to achieve what I mentioned in my last comment, you'd have to store the X, Y, and Z elements as separate components, e.g., PositionX, PositionY, PositionZ.  A bundle might make this a little more convenient, e.g., a PositionComponent.  Right now we have Transforms and GlobalTransforms, but from what I can tell, these are also laid out the same way that the Vec3 would be, which doesn't help with writing vectorized code.
